File: chapter-1/4-the-era-of-multithreading-and-parallelism/c++/thread.cpp
#include <iostream>
#include <thread>
#include <vector>

void process_item(int item) {
	std::cout << "Processed: " << item * 2 << std::endl;
}

int main() {
	std::vector<int> data = {1, 2, 3, 4, 5};
	std::vector<std::thread> threads;

	for (const auto &item : data) {
		threads.push_back(std::thread(process_item, item));
	}

	for (auto &thread : threads) {
		thread.join();
	}

	return 0;
}

// Processed: Processed: 26

// Processed: 4
// Processed: 8
// Processed: 10

File: chapter-1/4-the-era-of-multithreading-and-parallelism/rust/thread.rs
use std::thread;

fn main() {
	let data = vec![1, 2, 3, 4, 5];
	let mut handles = vec![];

	for &item in &data {
		handles.push(thread::spawn(move || {
			println!("Processed: {}", item * 2);
		}));
	}

	for handle in handles {
		handle.join().unwrap();
	}

}

//  rustc thread.rs && ./thread

// Processed: 2
// Processed: 6
// Processed: 4
// Processed: 8
// Processed: 10
